### PR TITLE
Refactor Connect to web components

### LINKED ISSUE #320: Rewrite Connect UI in non-React way
Stencil.js is a build tool that uses JSX to create web components that can be used in any front-end framework. Stenciljs is from the team of ionic, they made it so that they are less dependend on angular.

https://stenciljs.com/

This will make it more appealing to a lot more developers.

### LINKED ISSUE #587: [connect] dependency errors
Using the latest connect, it does not automatically install `@blockstack/stacks-transactions`

```
error - ./node_modules/@blockstack/connect/dist/connect.cjs.development.js:12:0
Module not found: Can't resolve '@blockstack/stacks-transactions'
```

### PR DESCRIPTION
**Pull request by hstove**
Our original implementation of Connect's UI was written in React. This was great for getting off the ground, because it allows us to re-use our internal UI library. However, importing Connect in non-React apps meant you now had to include React as a dependency. We also ran into complications that were caused by leaky CSS from the parent app.

We've talked about #320 for a while, and we decided it's time to take the plunge. Thanks to @jorishermans for suggesting that we use Stencil, which is a great solution for our needs.

This is a work in progress. Right now, the `connect-ui` is a new package, which is just the Web Components for Connect's UI. `connect` is now just the core Connect APIs, and a helper function to show UI.

Current work is related to refactoring what is now `connect-react` - I need to strip out all the React code in `connect-react` related to the Connect modal, which is re-used inside the authenticator, so we'll move that code into `app`. Finally, I'll refactor the remaining React hooks to continue to exist, but to use the Web Components for UI.

### CODE REVIEW

File: packages/connect-ui/src/components/modal/modal.sass [hunk 1]
```diff
@@ -0,0 +1,252 @@
+.modal-container {
```
Comment (kyranjamie) [inline packages/connect-ui/src/components/modal/modal.sass]: File should be `.scss` right? .sass would be sans-semicolons

File: packages/connect-ui/src/components/modal/modal.tsx [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+  styleUrl: 'modal.sass',
+  assetsDirs: ['screens', 'assets'],
+})
+export class Modal {
+  @Prop() authOptions: AuthOptions;
+  @Event()
+  signUp: EventEmitter;
+
+  @Event()
+  signIn: EventEmitter;
```
Comment (kyranjamie) [inline packages/connect-ui/src/components/modal/modal.tsx]: I miss describing events like this

Comment (hstove) [reply packages/connect-ui/src/components/modal/modal.tsx]: And it's all DOM events, which makes it easy to handle outside of the component. Stencil has a nice events API.

File: packages/connect-ui/src/components/modal/modal.sass (hunk 1 above)
Comment (hstove) [reply packages/connect-ui/src/components/modal/modal.sass]: It compiles, but yes. I personally prefer SASS so I'll likely just strip the semis. Long term, I think you and Thomas have better ideas around CSS-in-JS solutions - we might even be able to share the styled-system-esque solution that Thomas is working on.

**Comment by dantrevino** [comment]
@hstove if you're still looking for beta testers, I'm happy to try this on our upcoming update to hoverlook.

**Comment by hstove** [comment]
@dantrevino Yep, definitely would appreciate your help here. I'll let you know, and comment here, when I have something on NPM to beta test. Thanks!!

**Comment by hstove** [comment]
@timstackblock this still needs a code review, but I'd love for you to start testing this when possible.

What's changed: only the UI that is inside of the app, before you open the authenticator popup.

What to test:

- The UI should look the exact same as it does in production currently, i.e. on https://testnet-demo.blockstack.org
- Authentication should work as expected - sign up and sign in

@jasperjansz I'm also going to assign you for high-fidelity UI review. I followed the Figma, but there may be some tweaks that I missed.

File: packages/connect-ui/src/components/modal/modal.sass (hunk 1 above)
Comment (aulneau) [reply packages/connect-ui/src/components/modal/modal.sass]: Yes, I think almost certainly we can use `@stacks/ui-core` and `@stacks/ui-theme` here :) still WIP tho

**Comment by hstove** [comment]
Bug report from @njordhov -

> FYI: When attempting to integrate Connect stencil edition in a React app but without using its react features (hooks, context) I get an error when calling showBlockstackConnect saying connectUi.defineCustomElements is not a function. This happens whether the project depends on@blockstack/connect  or on@blockstack/connect-react with 2.12.2-alpha.41d15c8.0 as version as well as the previous 2.12.2-alpha.718aa48.0 version.

**Comment by hstove** [comment]
@markmhx I'm going to assign you to do a visual QA pass, hope that's OK!

File: packages/connect/src/ui.ts [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+  element.authOptions = authOptions;
+  document.body.appendChild(element);
+  const finishedWrapper = (finishedData: FinishedData) => {
+    element.remove();
+    const callback = authOptions.onFinish || authOptions.finished;
+    if (callback) {
+      callback(finishedData);
+    }
+  };
+  element.addEventListener('signUp', () => {
```
Comment (kyranjamie) [inline packages/connect/src/ui.ts]: These are custom events I take it, convention for native browser events is all lower case, any value keeping to that convention? 🤔

File: packages/connect/src/ui.ts [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+      onFinish: finishedWrapper,
+    });
+  });
+  const handleEsc = function (ev: KeyboardEvent) {
+    if (ev.key === 'Escape') {
+      document.removeEventListener('keydown', handleEsc);
+      element.remove();
+    }
+  };
+  element.addEventListener('closeModal', () => {
```
Comment (kyranjamie) [inline packages/connect/src/ui.ts]: Event might be better named `onModalClose` / `modalclose`

File: packages/connect-ui/src/index.html [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+      };
+      showBlockstackConnect(authOptions)
+    </script>
+    <style>
+      html, body { margin: 0px; }
+    </style>
+  </head>
+  <body>
+    <h1>App Content</h1>
+    <span>asdf</span>
```
Comment (kyranjamie) [inline packages/connect-ui/src/index.html:40]: ghjk

File: packages/connect-ui/connect/src/popup.js [hunk 6]
```diff
@@ -0,0 +1,92 @@
+// Width 2px wider than in-page dialog.
```
Comment (kyranjamie) [inline packages/connect-ui/connect/src/popup.js]: Could run prettier on this file/all .js

File: packages/app/src/components/app-icon/index.tsx [hunk 7]
```diff
@@ -6,7 +6,7 @@ import { Image } from '@components/image';
 import { AppState } from '@store';
 import { selectAppName, selectFullAppIcon } from '@store/onboarding/selectors';
 
-export const AppIcon = ({ ...rest }: BoxProps) => {
+export const AppIcon: React.FC<BoxProps> = ({ ...rest }: BoxProps) => {
```
Comment (kyranjamie) [inline packages/app/src/components/app-icon/index.tsx]:
```suggestion
export const AppIcon: React.FC<BoxProps> = ({ ...rest }) => {
```

File: packages/connect/src/ui.ts [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+    });
+  });
+  element.addEventListener('signIn', () => {
+    authenticate({
+      ...authOptions,
+      sendToSignIn: true,
+      onFinish: finishedWrapper,
+    });
+  });
+  const handleEsc = function (ev: KeyboardEvent) {
```
Comment (kyranjamie) [inline packages/connect/src/ui.ts]: Any reason this isn't an arrow fn?

File: packages/connect/src/ui.ts (hunk 4 above)
Comment (kyranjamie) [reply packages/connect/src/ui.ts]: Can pass an argument `@Event({ eventname: 'modalclose' })`

At a guess, consumers are more likely to listen for this event than emitting it, which might be a case for favoring past-tense event names over action/instruction names.

File: packages/connect/src/ui.ts (hunk 3 above)
Comment (kyranjamie) [reply packages/connect/src/ui.ts]: Do we need a prefix? Could this overlap with any other page events?

**Comment by markmhendrickson** [comment]
App icon in the top navigation of the Blockstack App is appearing too big:

Screen Shot 2020-09-11 at 17 03 21
Screen Shot 2020-09-11 at 17 05 42

**Comment by markmhendrickson** [comment]
The text in the introductory modal still appears to be affected by app styles. Here the header is left aligned instead of centered and the font weight generally looks bolder than it should:

Screen Shot 2020-09-11 at 17 04 25

Comparing to [Figma](https://www.figma.com/file/aXnNi8MOmN3CDrVwaDXXuqX3/Authenticator?node-id=5721%3A12547):

Screen Shot 2020-09-11 at 17 09 47

**Comment by markmhendrickson** [comment]
This link in the "How it works" modal appears smaller than surrounding text:

Screen Shot 2020-09-11 at 17 04 56

**Comment by markmhendrickson** [comment]
Let's capitalize "Secret Key" here since it's a proper name for our platform:

Screen Shot 2020-09-11 at 17 03 50

**Comment by markmhendrickson** [comment]
This is a bit of an earmark but the username field is still spellchecking and autosuggesting with email addresses in Chrome 😬:  

Screen Shot 2020-09-11 at 17 06 39

**Comment by markmhendrickson** [comment]
The "Install extension" link doesn't go anywhere: 

Screen Shot 2020-09-11 at 17 16 58

**Comment by markmhendrickson** [comment]
The modal bleeds off the viewport with smaller / mobile widths:

Screen Shot 2020-09-11 at 17 17 47
Screen Shot 2020-09-11 at 17 18 19

**Comment by hstove** [comment]
@markmhx thanks for reviewing! I've just pushed fixes for everything you mentioned.

File: packages/connect/src/ui.ts (hunk 4 above)
Comment (hstove) [reply packages/connect/src/ui.ts]: Good suggestions - I went with `on` prefixes.

**Comment by hstove** [comment]
@markmhx have you had a chance to confirm the visual QA fixes?

**Comment by markmhendrickson** [comment]
Ah no, but I will now

**Comment by markmhendrickson** [comment]
To be clear, we still want @timstackblock to give this a full QA pass as well, correct?

**Comment by hstove** [comment]
Yes, definitely need a proper QA on this as well. It touches both the in-app modal as well as the popup onboarding UI.

**Comment by markmhendrickson** [comment]
@timstackblock let's make sure to cross-reference all issues you create while QA'ing this PR, such as the ones I see you've created here: https://github.com/blockstack/ux/issues/604 https://github.com/blockstack/ux/issues/605 https://github.com/blockstack/ux/issues/606

**Comment by markmhendrickson** [comment]
@hstove I'm not seeing any fixes on the staging app as listed here: https://github.com/blockstack/ux/pull/582#issuecomment-685983677

It looks perhaps like redeployment may have been skipped for some reason?

Screen Shot 2020-09-25 at 14 17 18

**Comment by hstove** [comment]
> It looks perhaps like redeployment may have been skipped for some reason?

Late response @markmhx , but those tasks never run in PRs. They are for production deploys (to Vercel and NPM). We have different jobs, mainly `Deploy blockstack apps`, which deploy to staging apps.

**Pull request merged**